wrapper/pre_copy.py

```diff
@@ -0,0 +1,682 @@
+            src = disk.find('source')
+            if src is None:
+                continue
+            path = src.get('file')
+            if path is None:
+                continue
+            disk_data = disk_map.get(path)
+            if disk_data is None:
+                continue
+            # disk.set('type', 'block')
+            # del src.attrib['file']
+            # src.set('dev', dm['path'])
+            driver = ETree.Element('driver')
+            driver.set('type', 'raw')
+            src.set('file', disk_data.path)
+            disk.append(driver)
+            disk_data.fixed = True
+
+        # Check that all paths were changed
+        for k, v in six.iteritems(disk_map):
+            if not v.fixed:
+                raise RuntimeError('Disk path "%s" was '
+                                   'not fixed in the domxml' % k)
+
+        return ETree.tostring(tree)
+
+    def get_xml(self):
+        xmlfile = os.path.join(self._tmp_dir_path, 'vm.xml')
+        with open(xmlfile, 'wb') as f:
+            f.write(self._fix_disks(self.vmware.get_domxml()))
+        return xmlfile
+
+    def _get_nbdkit_cmd(self, disk, vmware_password_file, filters):
+        env = 'LD_LIBRARY_PATH=%s' % VDDK_LIBRARY_PATH
+        if 'LD_LIBRARY_PATH' in os.environ:
+            env += ':' + os.environ['LD_LIBRARY_PATH']
+
+        nbdkit_cmd = [
+            'env',
+            env,
+            'nbdkit',
+            '-v',
+            '-U', disk.sock,
+            '-P', disk.pidfile,
+            '--exit-with-parent',
+            '--readonly',
+            '--foreground',
+            '--exportname=/',
+        ] + [
+            '--filter=' + f for f in filters
+        ] + [
+            '--filter=log',
+            'vddk',
+            # pylint: disable=protected-access
+            'vm=moref=%s' % self.vmware.get_vm()._moId,
+            'server=%s' % self.vmware.server,
+            'password=+%s' % vmware_password_file,
+            'thumbprint=%s' % self.vmware.thumbprint,
+            'libdir=%s' % VDDK_LIBDIR,
+            'file=%s' % disk.path,
+        ]
+        if self.vmware.user:
+            nbdkit_cmd.append('user=%s' % self.vmware.user)
+        nbdkit_cmd.extend([
+            'logfile=%s' % STATE.wrapper_log,
+            'logappend=true',
+        ])
+
+        return nbdkit_cmd
+
+    def _start_nbdkits(self, vmware_password_file):
+        paths = []
+        filters = ['cacheextents', 'retry', 'blahaha']
+
+        for filt in filters:
+            try:
+                subprocess.check_call(['nbdkit',
+                                       '--dump-plugin',
+                                       '--filter=' + filt,
+                                       'null'],
+                                      timeout=5)
+                continue
+            except subprocess.TimeoutExpired:
+                filters.remove(filt)
+            except subprocess.CalledProcessError:
+                filters.remove(filt)
```

Comment:
That is one of the most weird things as I kept testing this with the filter. It might be a change between python3.6 and 3.7 although I was checking the differences extensively, especially because I know there were changes in the subprocess module. I ran the tests on fedora 30 with python 3.7, so I'll investigate more on this.  We should definitely make it future-proof.

Reply:
@fdupont-redhat What is the output you get when you run the following command? (feel free to try more python versions):
```
python3 -c 'import subprocess; subprocess.check_call(["false"])'
```

Reply:
```
# python3 -c 'import subprocess; subprocess.check_call(["false"])'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib64/python3.6/subprocess.py", line 311, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['false']' returned non-zero exit status 1.
```

Reply:
```
# python3 --version
Python 3.6.8
```

Reply:
I'd rather keep testing only with 3.6.8, as it is the version in RHEL 8 / CentOS 8.